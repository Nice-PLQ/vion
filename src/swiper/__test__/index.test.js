import { mount } from '@vue/test-utils';
import { triggerDrag, next } from '../../../test';
import VSwiper from '../swiper.vue';
import VSwiperItem from '../../swiper-item';

// offsetWidth mock
const originalOffsetWidth = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetWidth');
beforeAll(() => {
  Object.defineProperty(HTMLElement.prototype, 'offsetWidth', { configurable: true, value: 343 });
});

afterAll(() => {
  Object.defineProperty(HTMLElement.prototype, 'offsetWidth', originalOffsetWidth);
});

describe('Swiper', () => {
  test('渲染', async () => {
    const fn = jest.fn();
    const wrapper = mount({
      template: `
        <v-swiper @change="change" :show-indicators="false">
          <v-swiper-item style="height: 150px;" v-for="n in 4" :key="n">{{n}}</v-swiper-item>
        </v-swiper>
      `,
      methods: {
        change: fn
      }
    }, {
      global: {
        components: { VSwiper, VSwiperItem }
      }
    });
    const el = wrapper.find('.v-swiper__list');
    triggerDrag(el, -100, 0);
    await next();
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('.v-swiper__indicator').exists()).toBe(false);
    expect(el.element.style.transform).toMatch(/-686px/);

    el.trigger('transitionend');
    expect(fn).toHaveBeenCalledWith(1);
  });

  test('自动轮播', async () => {
    const wrapper = mount({
      template: `
        <v-swiper auto-play :interval="100">
          <v-swiper-item style="height: 150px;" v-for="n in 4" :key="n">{{n}}</v-swiper-item>
        </v-swiper>
      `
    }, {
      global: {
        components: { VSwiper, VSwiperItem }
      }
    });
    const el = wrapper.find('.v-swiper__list');
    triggerDrag(el, -5, 0);
    el.trigger('transitionend');
    await next(300);
    expect(el.element.style.transform).toMatch(/-686px/);
  });

  test('向右滑动后展示最后一个', async () => {
    const fn = jest.fn();
    const wrapper = mount({
      template: `
        <v-swiper @change="change">
          <v-swiper-item style="height: 150px;" v-for="n in 4" :key="n">{{n}}</v-swiper-item>
        </v-swiper>
      `,
      methods: {
        change: fn
      }
    }, {
      global: {
        components: { VSwiper, VSwiperItem }
      }
    });
    const el = wrapper.find('.v-swiper__list');
    triggerDrag(el, 100, 0);
    expect(fn).toHaveBeenCalledWith(1);
    expect(fn).toHaveBeenCalledWith(4);
  });
});
