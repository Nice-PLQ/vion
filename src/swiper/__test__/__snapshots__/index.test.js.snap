// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swiper 渲染 1`] = `
<div
  class="v-swiper"
>
  <ul
    class="v-swiper__list"
    style="transform: translate3d(-686px, 0, 0); transition: transform 500ms;"
  >
    
    <li
      class="v-swiper__item"
    >
      <div
        class="v-swiper__item-inner"
        style="height: 150px;"
      >
        
        4
        
      </div>
    </li>
    <li
      class="v-swiper__item"
    >
      <div
        class="v-swiper__item-inner"
        style="height: 150px;"
      >
        
        1
        
      </div>
    </li>
    <li
      class="v-swiper__item"
    >
      <div
        class="v-swiper__item-inner"
        style="height: 150px;"
      >
        
        2
        
      </div>
    </li>
    <li
      class="v-swiper__item"
    >
      <div
        class="v-swiper__item-inner"
        style="height: 150px;"
      >
        
        3
        
      </div>
    </li>
    <li
      class="v-swiper__item"
    >
      <div
        class="v-swiper__item-inner"
        style="height: 150px;"
      >
        
        4
        
      </div>
    </li>
    <li
      class="v-swiper__item"
    >
      <div
        class="v-swiper__item-inner"
        style="height: 150px;"
      >
        
        1
        
      </div>
    </li>
    
  </ul>
  <!--v-if-->
</div>
`;
